### PR TITLE
(feat)Snapshot improve.

### LINKED ISSUE #276: Snapshot improve
### Your question

The current design of snapshot is to configure the snapshot frequency. There is a case where only a few logs are written in the two snapshot intervals. Is it necessary to take another snapshot at this time?

### Your advice

Maybe we can provide an option: `distance`
Snapshot is executed when `fsm.getLastAppliedIndex - lastSnapshotId > distance`, otherwise it returns directly

### Environment

- SOFAJRaft version:
- JVM version (e.g. `java -version`):
- OS version (e.g. `uname -a`):
- Maven version:
- IDE version:

### PR DESCRIPTION
**Pull request by zongtanghu**
### Motivation:

Snapshot improve to add a option variable dist for judging the distance between state machine lastAppliedIndex and lastSnapshotIndex.

### Modification:

(1)NodeOptions.java.
(2)SnapshotExecutorImpl.java 

### Result:

Fixes #276 

If there is no issue then describe the changes introduced by this PR.

### CODE REVIEW

**Comment by zongtanghu** [comment]
@fengjiachun @killme2008  please help to review this simple pr.Thanks.

**Comment by killme2008** [comment]
@fengjiachun  我们的 commit message 规范没有放到 contribute 文档？  commit 日志需要规范下。

File: jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java [hunk 1]
```diff
@@ -67,6 +67,13 @@
     // Default: 3600 (1 hour)
     private int                             snapshotIntervalSecs   = 3600;
 
+    // A snapshot saving would be triggered when state machine's lastAppliedIndex value
```
Comment (killme2008) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java]: 这个说明有问题，他和 snapshotIntervalSecs 是 and 关系，建议说明的更详细一些

File: jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java [hunk 2]
```diff
@@ -67,6 +67,13 @@
...
+    // minus lastSnapshotId value is greater than snapshotIntervalDist value.
+    // If |snapshotIntervalDist| <= 0, the distance based snapshot would be disable.
+    //
+    // Default: 0
+    private int                             snapshotIntervalDist   = 0;
```
Comment (killme2008) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java]: 还是 Distance 吧，不需要省略这么几个字母， dist 容易误导。并且我觉的命名 `snapshotLogIndexMargin` 似乎更合适，跟 `NodeOptions#catchupMargin` 类似的语义

File: jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java [hunk 3]
```diff
@@ -344,9 +359,10 @@ public NodeOptions copy() {
     @Override
```
Comment (killme2008) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java]: 遗漏实现 copy 方法

**Comment by fengjiachun** [comment]
> @fengjiachun 我们的 commit message 规范没有放到 contribute 文档？ commit 日志需要规范下。

README.md 中有链接 [How to contribute](https://www.sofastack.tech/projects/sofa-jraft/how-to-contribute-code-to-sofajraft/ )

```
git commit -am '(feat) 添加xx功能'
git commit -am '(fix) 修复xx功能'
```

File: jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java (hunk 1 above)
Comment (zongtanghu) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java]: Okay, I will the code comments.

File: jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java (hunk 3 above)
Comment (zongtanghu) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java]: Okay  I will add the codes in copy method.

File: jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java (hunk 2 above)
Comment (zongtanghu) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/option/NodeOptions.java]: > > @fengjiachun 我们的 commit message 规范没有放到 contribute 文档？ commit 日志需要规范下。

Okay, I will observe the commit rules.

**Comment by zongtanghu** [comment]
I adjust my codes ,and push the codes again.Please help to review my codes.Thanks. @killme2008 @fengjiachun @masaimu

**Comment by fengjiachun** [comment]
@killme2008 can we merge this pr now？

File: jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java [hunk 4]
```diff
@@ -324,6 +324,17 @@ public void doSnapshot(final Closure done) {
                 Utils.runClosureInThread(done);
                 return;
             }
+
+            if (this.fsmCaller.getLastAppliedIndex() - this.lastSnapshotIndex < this.node.getOptions()
+                .getSnapshotLogIndexMargin()) {
+                // If state machine's lastAppliedIndex value minus lastSnapshotId value is
+                // less than or equal snapshotIntervalDist value, then directly return.
+                doUnlock = false;
```
Comment (killme2008) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java]: @zongtanghu  Can you add a debug log here?  Log the margin and says that ignore this snapshot time by snapshotLogIndexMargin setting.

Comment (zongtanghu) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java]: Okay, I will add the debug log that you said.

File: jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java [hunk 5]
```diff
@@ -329,6 +329,9 @@ public void doSnapshot(final Closure done) {
                 .getSnapshotLogIndexMargin()) {
                 // If state machine's lastAppliedIndex value minus lastSnapshotId value is
                 // less than or equal snapshotIntervalDist value, then directly return.
+                LOG.debug(
+                    "SnapshotLogIndexMargin={} , ignore this time of snapshot by snapshotLogIndexMargin setting",
+                    this.node.getOptions().getSnapshotLogIndexMargin());
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java]: 日志在明确一些，哪个 node？ 另外 `SnapshotLogIndexMargin={} ,`多一个空格

**Comment by zongtanghu** [comment]
Okay I will adjust this log codes.



huzongtang@cmss.chinamobile.com
 
From: block
Date: 2019-12-11 13:16
To: sofastack/sofa-jraft
CC: Hu Zongtang; Mention
Subject: Re: [sofastack/sofa-jraft] (feat)Snapshot improve. (#365)
@fengjiachun commented on this pull request.


In jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java:
> @@ -329,6 +329,9 @@ public void doSnapshot(final Closure done) {
                 .getSnapshotLogIndexMargin()) {
                 // If state machine's lastAppliedIndex value minus lastSnapshotId value is
                 // less than or equal snapshotIntervalDist value, then directly return.
+                LOG.debug(
+                    "SnapshotLogIndexMargin={} , ignore this time of snapshot by snapshotLogIndexMargin setting",
+                    this.node.getOptions().getSnapshotLogIndexMargin());

日志在明确一些，哪个 node？ 另外 SnapshotLogIndexMargin={} ,多一个空格
―

File: jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java [hunk 6]
```diff
@@ -324,6 +324,22 @@ public void doSnapshot(final Closure done) {
...
                 return;
             }
+
+            if (this.fsmCaller.getLastAppliedIndex() - this.lastSnapshotIndex < this.node.getOptions()
+                .getSnapshotLogIndexMargin()) {
+                // If state machine's lastAppliedIndex value minus lastSnapshotId value is
+                // less than or equal snapshotIntervalDist value, then directly return.
+                if (this.node != null) {
+                    LOG.debug(
+                        "Node {} snapshotLogIndexMargin={}, ignore this time of snapshot by snapshotLogIndexMargin setting.",
```
Comment (killme2008) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java]: 可以把实际的差值也打印下

Comment (fengjiachun) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/storage/snapshot/SnapshotExecutorImpl.java]: 被我手快合并了，@zongtanghu 再提一个 pr ？

**Pull request merged**